### PR TITLE
Replace `sha256` with `blake3` hash in image-optimizer

### PR DESCRIPTION
**Pull request by Brooooooklyn**
Here is benchmark suite for the `image-optimizer` scenario: https://github.com/Brooooooklyn/blake-hash/blob/main/bench-next.mjs

And the result shows `Blake3` hasher is much more faster than `sha256`:

### Calculate hash for `[3, 'http://abc.xyz/logo.webp', 256, 0.9, 'image/webp']`

```text
Running "digest hash into url-safe-base64" suite...
Progress: 100%

  blake3:
    1 512 961 ops/s, ±1.87%   | fastest

  blake2b:
    276 321 ops/s, ±2.97%     | slowest, 81.74% slower

  blake2s:
    290 832 ops/s, ±3.23%     | 80.78% slower

  sha256:
    296 197 ops/s, ±4.83%     | 80.42% slower

Finished 4 cases!
  Fastest: blake3
  Slowest: blake2b
```

### Calculate hash for a [png file](https://github.com/Brooooooklyn/blake-hash/blob/main/anime-girl.png)

```text
Running "digest big file" suite...
Progress: 100%

  blake3:
    6 729 ops/s, ±1.14%   | fastest

  blake2b:
    1 240 ops/s, ±0.61%   | 81.57% slower

  blake2bp:
    2 251 ops/s, ±2.40%   | 66.55% slower

  blake2s:
    722 ops/s, ±1.27%     | 89.27% slower

  blake2sp:
    2 417 ops/s, ±1.61%   | 64.08% slower

  sha256:
    468 ops/s, ±1.43%     | slowest, 93.05% slower

Finished 6 cases!
  Fastest: blake3
  Slowest: sha256
```

### CODE REVIEW

File: packages/next/package.json [hunk 1]
```diff
@@ -67,6 +67,7 @@
   "dependencies": {
     "@babel/runtime": "7.15.4",
     "@hapi/accept": "5.0.2",
+    "@napi-rs/blake-hash": "1.2.0",
```
Comment (styfle) [inline packages/next/package.json:70]: Looks like this is going to add an additional 1+ MB https://packagephobia.com/result?p=%40napi-rs%2Fblake-hash

File: packages/next/server/image-optimizer.ts [hunk 2]
```diff
@@ -584,18 +586,6 @@ function getSupportedMimeType(options: string[], accept = ''): string {
...
-  const hash = createHash('sha256')
-  for (let item of items) {
-    if (typeof item === 'number') hash.update(String(item))
-    else {
-      hash.update(item)
-    }
-  }
-  // See https://en.wikipedia.org/wiki/Base64#Filenames
-  return hash.digest('base64').replace(/\//g, '-')
-}
```
Comment (styfle) [inline packages/next/server/image-optimizer.ts:597]: This change is likely semver major because it will invalidate the existing cache for all images

**Review by styfle — COMMENTED** [review]
Did you compare the performance to any other built-in hashes?

I would prefer to avoid adding another dependency, especially if it means we're going to invalidate the cache for all Next.js deployments.

**Comment by Brooooooklyn** [comment]
> 
> 
> Did you compare the performance to any other built-in hashes?
> 
> I would prefer to avoid adding another dependency, especially if it means we're going to invalidate the cache for all Next.js deployments.

From this article: https://medium.com/asecuritysite-when-bob-met-alice/blake3-3716708235ac, `Blake3` is significantly faster than the other security hash algorithm.

**Comment by styfle** [comment]
Whats the digest size? Is this still 256 bit?

**Comment by Brooooooklyn** [comment]
> 
> 
> Whats the digest size? Is this still 256 bit?

Yes, `blake3` is 256 bit https://en.wikipedia.org/wiki/BLAKE_(hash_function)#BLAKE3

**Comment by creationix** [comment]
Assuming I understand the use case correctly here, I suggest we also compare non-cryptographic hashes.  In particular, I recommend adding the native and wasm versions and seeing how they compare.  I'm sure the second one is faster, but the first may be plenty fast and it's less headache being webassembly instead of native.

- https://www.npmjs.com/package/xxhash-wasm
- https://www.npmjs.com/package/xxhash-addon

As far as the chances of collisions I understand there are two keyspaces in which we want to avoid collisions:
- Each application has a local .next folder and we want all unique images to have unique hashes or the wrong image will be shown sometimes.  This should stay well below 1,000,000 unique values and is scoped to a single security domain (attacks would be against self)
- Browsers use the etag to know when to reuse a value in their cache.  This is scoped even more narrowly since it's per url.

Assuming a high quality non-cryptographic 64 bit hash the most likely collision is in the .next folder and it's about a 1-in-50,000,000 chance of there being a single collision among all of 1,000,000 unique images in that single application.  I consider this plenty safe considering the worse case is that 1 in 50 million mega sites may have one image rendered wrong.  A more likely scenario is a site with 2,000 unique images which has a 1-in-10,000,000,000,000 change of a single collision.

The etag is a non issue since there are unlikely to ever be more than two images for the same url in the browser cache so the chances of that collision are more like 1-in-18,000,000,000,000,000,000

**Comment by creationix** [comment]
If we really want to stick to built-in hashes in openssl and really care about performance, then maybe consider blake2 as it's pretty fast compared to sha256 and has a variable hash size as a supported feature so you could chose 64 bits and still be safe enough from collisions.

**Comment by Brooooooklyn** [comment]
I will add [xxhash](https://github.com/napi-rs/node-rs/blob/main/packages/xxhash/README.md) to benchmark tomorrow

File: packages/next/package.json (hunk 1 above)
Comment (styfle) [reply packages/next/package.json:70]: I checked and its only 670KB on macOS so I'm guessing this is suffering from the [same linux bug](https://twitter.com/styfle/status/1458115734496759816) we saw before where both musl and gnu are installed.

**Comment by timneutkens** [comment]
@Brooooooklyn just as a FYI webpack has xxhash64 using wasm in the latest version as an opt-in option and is enabled in Next.js

**Comment by Brooooooklyn** [comment]
### Benchmark result with `xxhash` and `node:blake2b512`

<details>
<summary>Calculate hash for a [3, 'http://abc.xyz/logo.webp', 256, 0.9, 'image/webp']
</summary>
<p>

```text
Running "digest hash into url-safe-base64" suite...
  blake3:
    1 730 591 ops/s, ±1.64%   | fastest

  blake2b:
    288 755 ops/s, ±2.21%     | slowest, 83.31% slower

  blake2s:
    292 547 ops/s, ±2.89%     | 83.1% slower

  node:sha256:
    324 604 ops/s, ±4.30%     | 81.24% slower

  node:blake2b512:
    301 157 ops/s, ±2.49%     | 82.6% slower

  xxh3:
    569 721 ops/s, ±1.44%     | 67.08% slower

  xxh64:
    544 897 ops/s, ±1.84%     | 68.51% slower

Finished 7 cases!
  Fastest: blake3
  Slowest: blake2b
```

</p>
</details>  


<details>
<summary>Calculate hash for 888kb PNG file</summary>
<p>

```text
Running "digest big file" suite...
  blake3:
    6 814 ops/s, ±1.13%    | 60.93% slower

  blake2b:
    1 261 ops/s, ±0.34%    | 92.77% slower

  blake2bp:
    2 329 ops/s, ±1.32%    | 86.65% slower

  blake2s:
    745 ops/s, ±0.44%      | 95.73% slower

  blake2sp:
    2 477 ops/s, ±1.08%    | 85.8% slower

  node:sha256:
    470 ops/s, ±1.12%      | slowest, 97.31% slower

  node:blake2b512:
    826 ops/s, ±1.64%      | 95.26% slower

  xxh64:
    13 583 ops/s, ±1.97%   | 22.12% slower

  xxh3:
    17 441 ops/s, ±1.30%   | fastest

Finished 9 cases!
  Fastest: xxh3
  Slowest: node:sha256
```

</p>
</details> 


`xxh3` is ***almost 3x faster*** than `blake3` in the **Big Image file**  scenario, and more than ***3x slower*** than `blake3`.

The `node:blake2b512` is almost the same performance with `node:sha256` in the small input scenario and ***2x faster*** in the big file input scenario.

> Assuming a high quality non-cryptographic 64 bit hash, the most likely collision is in the .next folder and it's about a 1-in-50,000,000 chance of there being a single collision among all of 1,000,000 unique images in that single application. I consider this plenty safe considering the worse case is that 1 in 50 million mega sites may have one image rendered wrong. 

I agree with it. 

If non-cryptographic 64 bit hash is a choice, I'll recommend https://github.com/tkaitchuck/ahash which is even faster than `xxhash`, then I'll develop Node.js binding for it.

If the security hash algorithm is still chosen, `Blake3` is the best at this moment.

**Comment by Brooooooklyn** [comment]
> @Brooooooklyn just as a FYI webpack has xxhash64 using wasm in the latest version as an opt-in option and is enabled in Next.js

I've been aware of this. The benchmark in [xxhash](https://github.com/napi-rs/node-rs/blob/main/packages/xxhash/README.md) is compare the native version to `webpack` verison: https://github.com/napi-rs/node-rs/blob/main/packages/xxhash/benchmark/xxhash.js#L6. From the result, `wasm` is almost  `30%` slower  than the native version.

**Comment by styfle** [comment]
I like using the same algorithm as webpack so I would vote for xxhash. Ideally we could stick to one algorithm and one package for all file hashing, not just images here.

Do you agree @timneutkens @creationix ? Also curious if we can replace this in other places.

**Comment by creationix** [comment]
I also like using xxhash since we use it other places as well already.

**Review by ijjk — COMMENTED** [review]
Hi, looks like from the above discussion we want to converge on leveraging `xxhash` instead so I'm gonna close this PR for now, thanks for looking into this!

**Pull request closed without merging**